### PR TITLE
TEET-1676 Initial implementation: Adding materials or products to assets (cost items) 

### PR DESCRIPTION
**Pull request by outamaa**
This PR implements the basic CRUD operations for materials. Materials can only for components that are leafs in the asset type hierarchy.

What's _not_ implemented:
- Cost group functionality for materials
- Adding materials with no attributes (I assume this there are none as they get specified)

### CODE REVIEW

File: app/frontend/src/cljs/teet/asset/cost_items_view.cljs [hunk 1]
```diff
@@ -1,41 +1,41 @@
...
-            [teet.asset.cost-items-map-view :as cost-items-map-view]))
+            [teet.ui.query :as query]
+            [teet.ui.select :as select]
+            [teet.ui.table :as table]
+            [teet.ui.text-field :as text-field]
+            [teet.ui.typography :as typography]
+            [teet.ui.url :as url]
+            [teet.user.user-model :as user-model]
+            [teet.util.collection :as cu]
+            [teet.util.datomic :as du]))
```
Comment (outamaa) [inline app/frontend/src/cljs/teet/asset/cost_items_view.cljs:38]: Just sorted the namespaces.

File: app/frontend/src/cljs/teet/asset/cost_items_view.cljs [hunk 2]
```diff
@@ -456,27 +456,53 @@
...
-                :action (r/partial delete-component! (:db/id c))}
+                :action (r/partial delete-fn (:db/id c))}
                (tr [:buttons :delete])]])]]
          [component-rows {:e! e!
                           :locked? locked?
                           :components (:component/components c)
                           :level (inc (or level 0))}]]))]))
 
-(defn- components-tree*
+(defn- children-tree*
```
Comment (outamaa) [inline app/frontend/src/cljs/teet/asset/cost_items_view.cljs:466]: Small changes to accommodate materials as well as components.

File: app/frontend/src/cljs/teet/asset/cost_items_view.cljs [hunk 3]
```diff
@@ -622,6 +660,60 @@
...
+        ;; Attributes for material
+        [form-paper
+         [:<>
+          (label material-type)
+          [attributes {:e! e!
+                       :attributes (some-> material-type :attribute/_parent)
+                       :inherits-location? true
+                       :material-oid material-oid
+                       :cost-item-data cost-item-data
+                       :common? false
```
Comment (outamaa) [inline app/frontend/src/cljs/teet/asset/cost_items_view.cljs:699]: How do the new common types affect this?

File: app/backend/src/clj/teet/asset/asset_db.clj [hunk 4]
```diff
@@ -511,3 +541,19 @@
...
       (d/as-of db c))))
+
+(defn leaf-component?
+  "Is the component a leaf component? A component is a leaf component if
+  the component type does not allow child components."
+  [db component-id]
+  ;; Does there exist a ?ctype such that
+  ;; - ?ctype is the ctype of component-id and
+  ;; - ?ctype is no other ctype's parent
+  (->> (d/q '[:find (some? ?ctype)
```
Comment (tatut) [inline app/backend/src/clj/teet/asset/asset_db.clj:552]: can you use `some?` as an aggregate? it's not documented in the list of builtin aggregates

**Pull request merged**